app/vmselect/promql/exec.go

```diff
@@ -56,9 +167,11 @@ func Exec(ec *EvalConfig, q string, isFirstPointOnly bool) ([]netstorage.Result,
 	maySort := maySortResults(e, rv)
 	result, err := timeseriesToResult(rv, maySort)
 	if err != nil {
-		return nil, err
+		stopCh <- err
+		return
 	}
-	return result, err
+
+	resultCh <- result
```

Comment:
The goroutine blocks indefinitely at this line after the `CancelRunninQuery` is called - it puts an entry to `stopCh`, so the parent goroutine unblocks on the line 136, while `resultCh` is never read. This leads to goroutine leak.

Probably it is better to use a single `resultCh` for returning both results and errors, including `canceled query` error.